File: lib/src/features/member_staff/models/time_slot.dart
/// Model representing a time slot in a staff member's schedule.
class TimeSlot {
  final String? id;
  final String? staffId;
  final DateTime date;
  final String startTime;
  final String endTime;
  final bool isBooked;
  
  TimeSlot({
    this.id,
    this.staffId,
    required this.date,
    required this.startTime,
    required this.endTime,
    this.isBooked = true,
  });
  
  factory TimeSlot.fromJson(Map<String, dynamic> json) {
    return TimeSlot(
      id: json['id'],
      staffId: json['staff_id'],
      date: DateTime.parse(json['date']),
      startTime: json['start_time'],
      endTime: json['end_time'],
      isBooked: json['is_booked'] ?? true,
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      if (staffId != null) 'staff_id': staffId,
      'date': date.toIso8601String().split('T').first,
      'start_time': startTime,
      'end_time': endTime,
      'is_booked': isBooked,
    };
  }
  
  TimeSlot copyWith({
    String? id,
    String? staffId,
    DateTime? date,
    String? startTime,
    String? endTime,
    bool? isBooked,
  }) {
    return TimeSlot(
      id: id ?? this.id,
      staffId: staffId ?? this.staffId,
      date: date ?? this.date,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      isBooked: isBooked ?? this.isBooked,
    );
  }
  
  /// Checks if this time slot overlaps with another time slot.
  bool overlaps(TimeSlot other) {
    // Different dates don't overlap
    if (date.year != other.date.year || 
        date.month != other.date.month || 
        date.day != other.date.day) {
      return false;
    }
    
    // Convert times to minutes for easier comparison
    final thisStart = _timeToMinutes(startTime);
    final thisEnd = _timeToMinutes(endTime);
    final otherStart = _timeToMinutes(other.startTime);
    final otherEnd = _timeToMinutes(other.endTime);
    
    // Check for overlap
    return thisStart < otherEnd && thisEnd > otherStart;
  }
  
  /// Converts a time string (HH:MM) to minutes.
  int _timeToMinutes(String time) {
    final parts = time.split(':');
    return int.parse(parts[0]) * 60 + int.parse(parts[1]);
  }
}

File: backend-api/app/Models/TimeSlot.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class TimeSlot extends Model
{
    use HasFactory, SoftDeletes;

    /**
     * The table associated with the model.
     *
     * @var string
     */
    protected $table = 'time_slots';

    /**
     * The primary key for the model.
     *
     * @var string
     */
    protected $primaryKey = 'id';

    /**
     * Indicates if the IDs are auto-incrementing.
     *
     * @var bool
     */
    public $incrementing = false;

    /**
     * The data type of the auto-incrementing ID.
     *
     * @var string
     */
    protected $keyType = 'string';

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'staff_id',
        'date',
        'start_time',
        'end_time',
        'is_booked',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'date' => 'date',
        'is_booked' => 'boolean',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    /**
     * Get the staff that owns the time slot.
     */
    public function staff()
    {
        return $this->belongsTo(Staff::class);
    }

    /**
     * Check if this time slot overlaps with another time slot.
     *
     * @param  \App\Models\TimeSlot  $timeSlot
     * @return bool
     */
    public function overlaps(TimeSlot $timeSlot)
    {
        // Different dates don't overlap
        if (!$this->date->isSameDay($timeSlot->date)) {
            return false;
        }

        // Convert times to minutes for easier comparison
        $thisStart = $this->timeToMinutes($this->start_time);
        $thisEnd = $this->timeToMinutes($this->end_time);
        $otherStart = $this->timeToMinutes($timeSlot->start_time);
        $otherEnd = $this->timeToMinutes($timeSlot->end_time);

        // Check for overlap
        return $thisStart < $otherEnd && $thisEnd > $otherStart;
    }

    /**
     * Convert a time string (HH:MM) to minutes.
     *
     * @param  string  $time
     * @return int
     */
    private function timeToMinutes($time)
    {
        $parts = explode(':', $time);
        return (int) $parts[0] * 60 + (int) $parts[1];
    }

    /**
     * Scope a query to only include time slots for a specific date.
     *
     * @param  \Illuminate\Database\Eloquent\Builder  $query
     * @param  \Carbon\Carbon|string  $date
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function scopeForDate($query, $date)
    {
        return $query->whereDate('date', $date);
    }

    /**
     * Scope a query to only include booked time slots.
     *
     * @param  \Illuminate\Database\Eloquent\Builder  $query
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function scopeBooked($query)
    {
        return $query->where('is_booked', true);
    }

    /**
     * Scope a query to only include available time slots.
     *
     * @param  \Illuminate\Database\Eloquent\Builder  $query
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function scopeAvailable($query)
    {
        return $query->where('is_booked', false);
    }
}
